File: vortexdbg-api/src/main/java/com/vortexdbg/unix/struct/StdString.java
package com.vortexdbg.unix.struct;

import com.vortexdbg.Emulator;
import com.vortexdbg.pointer.VortexdbgStructure;
import com.sun.jna.Pointer;

import java.nio.charset.StandardCharsets;

public abstract class StdString extends VortexdbgStructure {

    public static StdString createStdString(Emulator<?> emulator, Pointer pointer) {
        if (emulator.is64Bit()) {
            return new StdString64(pointer);
        } else {
            return new StdString32(pointer);
        }
    }

    StdString(Pointer p) {
        super(p);
    }

    public final String getValue(Emulator<?> emulator) {
        return new String(getData(emulator), StandardCharsets.UTF_8);
    }

    public final byte[] getData(Emulator<?> emulator) {
        return getDataPointer(emulator).getByteArray(0, (int) getDataSize());
    }

    public abstract Pointer getDataPointer(Emulator<?> emulator);
    public abstract long getDataSize();

}

File: vortexdbg-api/src/main/kotlin/com/vortexdbg/unix/struct/StdString.kt
package com.vortexdbg.unix.struct

import com.vortexdbg.Emulator
import com.vortexdbg.pointer.VortexdbgStructure
import com.sun.jna.Pointer

import java.nio.charset.StandardCharsets

abstract class StdString internal constructor(p: Pointer?) : VortexdbgStructure(p) {

    fun getValue(emulator: Emulator<*>): String {
        return String(getData(emulator), StandardCharsets.UTF_8)
    }

    fun getData(emulator: Emulator<*>): ByteArray {
        return getDataPointer(emulator).getByteArray(0, getDataSize().toInt())
    }

    abstract fun getDataPointer(emulator: Emulator<*>): Pointer
    abstract fun getDataSize(): Long

    companion object {
        @JvmStatic
        fun createStdString(emulator: Emulator<*>, pointer: Pointer): StdString {
            return if (emulator.is64Bit()) {
                StdString64(pointer)
            } else {
                StdString32(pointer)
            }
        }
    }
}
